src/journal/writer.rs

```diff
@@ -260,37 +296,36 @@ impl Writer {
         seqno: u64,
     ) -> crate::Result<usize> {
         self.is_buffer_dirty = true;
+        self.buf.clear();
 
-        let mut hasher = xxhash_rust::xxh3::Xxh3::default();
-        let mut byte_count = 0;
+        let compression =
+            if self.compression_threshold > 0 && value.len() >= self.compression_threshold {
+                self.compression
+            } else {
+                CompressionType::None
+            };
 
-        self.buf.clear();
-        byte_count += self.write_start(1, seqno)?;
-        self.buf.clear();
+        self.buf.write_u8(Tag::SingleItem.into())?;
+        self.buf.write_u64::<LittleEndian>(seqno)?;
 
-        serialize_marker_item(
-            &mut self.buf,
+        let mut hashing_writer = HashingWriter::new(&mut self.buf);
+
+        serialize_item(
+            &mut hashing_writer,
             keyspace_id,
             key,
             value,
             value_type,
-            if self.compression_threshold > 0 && value.len() >= self.compression_threshold {
-                self.compression
-            } else {
-                CompressionType::None
-            },
+            compression,
         )?;
 
-        self.file.write_all(&self.buf)?;
+        let checksum = hashing_writer.finish();
 
-        hasher.update(&self.buf);
-        byte_count += self.buf.len();
+        self.buf.write_u64::<LittleEndian>(checksum)?;
 
-        self.buf.clear();
```

Comment:
I think originally the buf existed to calculate the checksum. With the streaming `HashingWriter` maybe the buf can be omitted completely, that way data is not written to twice.